### PR TITLE
Launch links in external application

### LINKED ISSUE #28: Allow opening links in external applications
### Is your feature request related to a problem? Please describe.

YouTube opens in browser. 

Spotify too.

That means I can't play the content very well.

### Describe the solution you'd like

Either by default or at least an option in the settings, allow to open any link not to lemmy in an external app.

### PR DESCRIPTION
**Pull request by FriederHannenheim**
Fixes #28

### CODE REVIEW

**Comment by zachatrocity** [comment]
Yay! You might win the first contribution badge. 

This seems like a really good candidate for a settings option. Would you be comfortable adding that?

**Comment by FriederHannenheim** [comment]
Sure

**Comment by FriederHannenheim** [comment]
@zachatrocity Okay I've implemented a setting toggle for this

**Comment by zachatrocity** [comment]
Great! Thanks. Please resolve conflicts.

@shocklateboy92 recently made some changes in this area so ping him if you have issues

File: lib/pages/settings/settings.dart [hunk 1]
```diff
@@ -195,6 +195,13 @@ class GeneralConfigPage extends StatelessWidget {
                 ),
               ),
             ),
+            SwitchListTile.adaptive(
+              title: const Text('Open links in external application'),
```
Comment (shocklateboy92) [inline lib/pages/settings/settings.dart:199]: Nit: Maybe "Always open links in external application"?

We do currently open non-http links in external apps. This just makes it unconditional.

File: lib/stores/config_store.dart [hunk 2]
```diff
@@ -49,6 +49,10 @@ abstract class _ConfigStore with Store {
   @JsonKey(defaultValue: false)
   bool amoledDarkMode = false;
 
+  @observable
+  @JsonKey(defaultValue: true)
+  bool openLinksInExternalApp = true;
```
Comment (shocklateboy92) [inline lib/stores/config_store.dart:54]: Maybe the default shouldn't be `true`.
The in-app view is already better with my changes. We'll continue to make it not suck.
Let's leave the users to enable this if/when they get frustrated.

File: lib/url_launcher.dart [hunk 3]
```diff
@@ -114,7 +116,11 @@ Future<bool> launchLink({
 }) async {
   final uri = Uri.tryParse(link);
   if (uri != null && await ul.canLaunchUrl(uri)) {
-    await ul.launchUrl(uri);
+    await ul.launchUrl(uri, mode: 
+        context.read<ConfigStore>().openLinksInExternalApp ?
+        ul.LaunchMode.externalApplication : 
+        ul.LaunchMode.inAppWebView
```
Comment (shocklateboy92) [inline lib/url_launcher.dart:122]: The false case should be `platformDefault`. There are cases where we shouldn't try to open in web view even without this setting.

**Review by shocklateboy92 — CHANGES_REQUESTED** [review]

File: lib/pages/settings/settings.dart (hunk 1 above)
Comment (bahuma20) [reply lib/pages/settings/settings.dart:199]: Maybe name it "Use in-app browser" and have it enabled by default?

**Review by bahuma20 — APPROVED** [review]

**Comment by FriederHannenheim** [comment]
Yes I am still working on it. I don't have too much time currently but I will try to get it done this evening

**Comment by FriederHannenheim** [comment]
@zachatrocity I believe the merge request is ready now. I rebased my changes on master and implemented all the requested changes

**Review by zachatrocity — APPROVED** [review]
LGTM 👊

**Comment by swmarks** [comment]
I've got a Samsung Fold 4 on Android 13

I hate to say it, but this completely breaks opening links for me unless I have my default browser set to Samsung Internet, whether the option is enabled or disabled in my account settings. No matter what post I try to open it says "Can't open in browser" at the bottom of the screen. I like Samsung Internet, but it's not my browser of choice at the moment. It would be good if anybody else can confirm if this is working or not working for them.

**Comment by bahuma20** [comment]
How can i install the latest build to test it?

**Comment by swmarks** [comment]
@bahuma20 I could potentially send you a built apk, but this isn't in a release yet

I found this ominous note when hovering over the canLaunchUrl method:

> On recent versions of Android and iOS, this will always return false unless the application has been configuration to allow querying the system for launch support. See [the README](https://pub.dev/packages/url_launcher#configuration) for details.

I think we need to remove that canLaunchUrl check for now. We can add some sort of sanity check back in later, but it's not working as is and people will have issues if we push this in a final release.

**Comment by swmarks** [comment]
I'm making another pull request for this.

**Comment by DraconicNEO** [comment]
This is currently not working for me, I'm using build 0.9.18 from the Play store and it's still opens all links with the inline browser. I can't see any options in the settings page to change this either.

**Comment by swmarks** [comment]
> This is currently not working for me, I'm using build 0.9.18 from the Play store and it's still opens all links with the inline browser. I can't see any options in the settings page to change this either.

Unfortunately, this merge didn't make it into that version release. We'll get it out the door soon, I'm sure.

**Comment by DraconicNEO** [comment]
> > This is currently not working for me, I'm using build 0.9.18 from the Play store and it's still opens all links with the inline browser. I can't see any options in the settings page to change this either.
> 
> Unfortunately, this merge didn't make it into that version release. We'll get it out the door soon, I'm sure.

Oh, that's unfortunate. Hopefully it'll be fixed soon.

**Pull request merged**